// @loaders.gl/polyfills/dist/es5/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ReadableStreamPolyfill", {
  enumerable: true,
  get: function get() {
    return _readableStreamPolyfill.ReadableStreamPolyfill;
  }
});
Object.defineProperty(exports, "BlobPolyfill", {
  enumerable: true,
  get: function get() {
    return _blobPolyfill.BlobPolyfill;
  }
});
Object.defineProperty(exports, "FileReaderPolyfill", {
  enumerable: true,
  get: function get() {
    return _fileReaderPolyfill.FileReaderPolyfill;
  }
});
Object.defineProperty(exports, "FilePolyfill", {
  enumerable: true,
  get: function get() {
    return _filePolyfill.FilePolyfill;
  }
});
Object.defineProperty(exports, "installFilePolyfills", {
  enumerable: true,
  get: function get() {
    return _polyfills.installFilePolyfills;
  }
});

var _globals = require("./utils/globals");

var _encoding = require("./libs/encoding");

var base64 = _interopRequireWildcard(require("./text-encoding/btoa.node"));

var _headers = _interopRequireDefault(require("./fetch-node/headers.node"));

var _response = _interopRequireDefault(require("./fetch-node/response.node"));

var _fetch = _interopRequireDefault(require("./fetch-node/fetch.node"));

var _encodeImage = require("./images-node/encode-image.node");

var _parseImage = require("./images-node/parse-image.node");

var _readableStreamPolyfill = require("./file/readable-stream-polyfill");

var _blobPolyfill = require("./file/blob-polyfill");

var _fileReaderPolyfill = require("./file/file-reader-polyfill");

var _filePolyfill = require("./file/file-polyfill");

var _polyfills = require("./file/polyfills");

var installTextEncoder = _encoding.TextEncoder && (!_globals.isBrowser || !('TextEncoder' in _globals.global));

if (installTextEncoder) {
  _globals.global['TextEncoder'] = _encoding.TextEncoder;
}

var installTextDecoder = _encoding.TextEncoder && (!_globals.isBrowser || !('TextDecoder' in _globals.global));

if (installTextDecoder) {
  _globals.global['TextDecoder'] = _encoding.TextDecoder;
}

if (!_globals.isBrowser && !('atob' in _globals.global) && base64.atob) {
  _globals.global['atob'] = base64.atob;
}

if (!_globals.isBrowser && !('btoa' in _globals.global) && base64.btoa) {
  _globals.global['btoa'] = base64.btoa;
}

if (!_globals.isBrowser && !('Headers' in _globals.global) && _headers["default"]) {
  _globals.global['Headers'] = _headers["default"];
}

if (!_globals.isBrowser && !('Response' in _globals.global) && _response["default"]) {
  _globals.global['Response'] = _response["default"];
}

if (!_globals.isBrowser && !('fetch' in _globals.global) && _fetch["default"]) {
  _globals.global['fetch'] = _fetch["default"];
}

if (!_globals.isBrowser && !('_encodeImageNode' in _globals.global) && _encodeImage.encodeImageNode) {
  _globals.global['_encodeImageNode'] = _encodeImage.encodeImageNode;
}

if (!_globals.isBrowser && !('_parseImageNode' in _globals.global) && _parseImage.parseImageNode) {
  _globals.global['_parseImageNode'] = _parseImage.parseImageNode;
}
