// @loaders.gl/polyfills/dist/es6/fetch-node/utils/stream-utils.node.js
import fs from 'fs';
import http from 'http';
import https from 'https';
import zlib from 'zlib';
import { toArrayBuffer } from './decode-data-uri.node';

const isRequestURL = url => url.startsWith('http:') || url.startsWith('https:');

export async function createReadStream(url, options) {
  if (!isRequestURL(url)) {
    const noqueryUrl = url.split('?')[0];
    return await new Promise((resolve, reject) => {
      const stream = fs.createReadStream(noqueryUrl, {
        encoding: null
      });
      stream.once('readable', () => resolve(stream));
      stream.on('error', error => reject(error));
    });
  }

  return await new Promise((resolve, reject) => {
    const requestFunction = url.startsWith('https:') ? https.request : http.request;
    const requestOptions = getRequestOptions(url, options);
    const req = requestFunction(requestOptions, res => resolve(res));
    req.on('error', error => reject(error));
    req.end();
  });
}
export function decompressReadStream(readStream, headers) {
  switch (headers.get('content-encoding')) {
    case 'br':
      return readStream.pipe(zlib.createBrotliDecompress());

    case 'gzip':
      return readStream.pipe(zlib.createGunzip());

    case 'deflate':
      return readStream.pipe(zlib.createDeflate());

    default:
      return readStream;
  }
}
export async function concatenateReadStream(readStream) {
  let arrayBuffer = new ArrayBuffer(0);
  return await new Promise((resolve, reject) => {
    readStream.on('error', error => reject(error));
    readStream.on('readable', () => readStream.read());
    readStream.on('data', chunk => {
      if (typeof chunk === 'string') {
        reject(new Error('Read stream not binary'));
      }

      const chunkAsArrayBuffer = toArrayBuffer(chunk);
      arrayBuffer = concatenateArrayBuffers(arrayBuffer, chunkAsArrayBuffer);
    });
    readStream.on('end', () => resolve(arrayBuffer));
  });
}

function getRequestOptions(url, options = {}) {
  const originalHeaders = options.headers || {};
  const headers = {};

  for (const key of Object.keys(originalHeaders)) {
    headers[key.toLowerCase()] = originalHeaders[key];
  }

  headers['accept-encoding'] = headers['accept-encoding'] || 'gzip,br,deflate';
  const urlObject = new URL(url);
  return {
    hostname: urlObject.hostname,
    path: urlObject.pathname,
    method: 'GET',
    ...options,
    ...(options.fetch || {}),
    headers
  };
}

function concatenateArrayBuffers(source1, source2) {
  const sourceArray1 = source1 instanceof ArrayBuffer ? new Uint8Array(source1) : source1;
  const sourceArray2 = source2 instanceof ArrayBuffer ? new Uint8Array(source2) : source2;
  const temp = new Uint8Array(sourceArray1.byteLength + sourceArray2.byteLength);
  temp.set(sourceArray1, 0);
  temp.set(sourceArray2, sourceArray1.byteLength);
  return temp.buffer;
}
